pkg/auth/providers/azure/clients/msclient/ms_graph_client.go

```diff
@@ -0,0 +1,199 @@
+package msclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/cache"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers/azure/clients"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+)
+
+type azureMSGraphClient struct {
+	authResult  *customAuthResult
+	userClient  *msgraph.UsersClient
+	groupClient *msgraph.GroupsClient
+}
+
+func (c azureMSGraphClient) MarshalTokenJSON() (string, error) {
+	b, err := json.Marshal(c.authResult.AccessToken)
+	return string(b), err
+}
+
+func (c azureMSGraphClient) GetUser(id string) (interface{}, error) {
+	u, _, err := c.userClient.Get(context.Background(), id, odata.Query{})
+	return u, err
+}
+
+func (c azureMSGraphClient) ListUsers(filter string) (interface{}, error) {
+	users, _, err := c.userClient.List(context.Background(), odata.Query{Filter: filter})
+	return users, err
+}
+
+func (c azureMSGraphClient) GetGroup(id string) (interface{}, error) {
+	g, _, err := c.groupClient.Get(context.Background(), id, odata.Query{})
+	return g, err
+}
+
+func (c azureMSGraphClient) ListGroups(filter string) (interface{}, error) {
+	groups, _, err := c.groupClient.List(context.Background(), odata.Query{Filter: filter})
+	return groups, err
+}
+
+func (c azureMSGraphClient) ListGroupMemberships(id string, _ interface{}) ([]string, error) {
+	groups, _, err := c.userClient.ListGroupMemberships(context.Background(), id, odata.Query{})
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, len(*groups))
+	for _, g := range *groups {
+		names = append(names, *g.ID)
+	}
+	return names, nil
+}
+
+type customAuthResult struct {
+	AccessToken    string    `json:"access_token,omitempty"`
+	ExpiresOn      time.Time `json:"expires_on,omitempty"`
+	GrantedScopes  []string  `json:"granted_scopes,omitempty"`
+	DeclinedScopes []string  `json:"declined_scopes,omitempty"`
```

Comment:
Can these be changed to camel case instead? That is what we use for json properties.